packages/hydrooj/src/handler/record.ts

```diff
@@ -338,13 +340,13 @@ class RecordMainConnectionHandler extends ConnectionHandler {
             if (!this.user.hasPerm(PERM.PERM_VIEW_PROBLEM)) pdoc = null;
         }
         if (this.applyProjection && typeof rdoc.input !== 'string') rdoc = contest.applyProjection(tdoc, rdoc, this.user);
-        if (this.pretest) this.send({ rdoc: omit(rdoc, ['code', 'input']) });
+        if (this.pretest) this.throttleSend(async () => ({ rdoc: omit(rdoc, ['code', 'input']) }));
         else {
-            this.send({
+            this.throttleSend(async () => ({
```

Comment:
Have to make sure that the last event of each submission was sent.

Reply:
That's how throttle works, right? It can be called multiple times but only the last call gets to be executed when `trailing` enabled.

Reply:
But this is different from RecordDetailHandler. RecordMainHandler sends updates of multiple submissions and you have to ensure the trailing of EACH one.

Reply:
Got it, added a queue to save the last message for each record identified by `rid`.